### PR TITLE
Add support for dune

### PR DESCRIPTION
**Pull request by Khady**
dune has the ability to format its files. Adding support through the `format-dune-file` command, even though it's [not perfect](https://github.com/ocaml/dune/issues/3516).

I can drop the ocamlformat change if necessary, or put it in a different PR. It was easier to have it when developing with a modern ocaml (4.14)

This work is sponsored by Ahrefs.

### CODE REVIEW

**Comment by hhugo** [comment]
Thanks.
I've opened https://github.com/hhugo/merge-fmt/pull/4 to upgrade the codebase to new versions of its deps.
Can you rebase you commit on top and add some expect test for the dune features ? (just look at other tests for inspiration)

**Comment by hhugo** [comment]
@Khady, you should be able to rebase on master and have the CI run

File: test/merge_dune.ml [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      [%expect
+        {|
+        File dune
+        (executable
+         (name aaaa)
+         (libraries unix)
+         (public_name pppp))
+
+        (library
+         (name liblib)) |}])
```
Comment (Khady) [inline test/merge_dune.ml]: the output is not correct, it results into `branch2` formatted and the content of `old_branch1` is dropped. IIUC the final state should be

```
(executable
 (name bbbb)
 (libraries unix)
 (public_name pppp))

(library
 (name liblib))

(alias
 (name runtest)
 (action
  (diff rebase.diff rebase.diff.gen)))
```

**Comment by Khady** [comment]
another problem when running merge_fmt by hand

```
$ _build/default/src/merge_fmt.exe 
merge-fmt: internal error, uncaught exception:
           Invalid_argument("Filename.chop_extension")
           Raised at Stdlib.invalid_arg in file "stdlib.ml", line 30, characters 20-45
           Called from Resolve_cmd.create_tmp in file "src/resolve_cmd.ml", line 71, characters 13-44
           Called from Resolve_cmd.fix in file "src/resolve_cmd.ml", line 84, characters 13-52
           Called from Resolve_cmd.resolve.(fun) in file "src/resolve_cmd.ml", line 122, characters 26-68
           Called from Base__Map.Accessors.iteri in file "src/map.ml" (inlined), line 1794, characters 19-40
           Called from Resolve_cmd.resolve in file "src/resolve_cmd.ml", line 116, characters 2-687
           Called from Cmdliner_term.app.(fun) in file "cmdliner_term.ml", line 24, characters 19-24
           Called from Cmdliner_eval.run_parser in file "cmdliner_eval.ml", line 34, characters 37-44
```

**Comment by Khady** [comment]
and now getting another error. I can't spend more time on it right now. Will try to come back to that in the coming weeks

using this test repo for reference https://github.com/Khady/merge-fmt-test
```
$ merge_fmt.exe --echo
+ git ls-files -u
+ git show c27aa6cb49462bf5414fce78a204491aebff17b9
+ git show 4df152dd4dba64fe71c148934c93831b3ff4a12f
+ git show cbc79883f33bbbde24480b79ac7592d9a01f8a4e
+ dune format-dune-file -- dune >  dune.ours
File "/tmp/test-git/dune", line 11, characters 17-18:
11 |  (libraries unix))
                      ^
Error: right parenthesis without matching left parenthesis
+ dune format-dune-file -- dune >  dune.theirs
File "/tmp/test-git/dune", line 11, characters 17-18:
11 |  (libraries unix))
                      ^
Error: right parenthesis without matching left parenthesis
+ dune format-dune-file -- dune >  dune.common
File "/tmp/test-git/dune", line 11, characters 17-18:
11 |  (libraries unix))
                      ^
Error: right parenthesis without matching left parenthesis
Failed to format dune.ours, dune.theirs, dune.common
Resolved 0/1 dune
+ git ls-files -u

```

**Comment by hhugo** [comment]
- I've rewrote the format command for dune files to make it more robust.
- I've updated the test, the previous version was failing to merge.

**Comment by hhugo** [comment]
Can you add a changelog entry and mark the PR as ready

**Comment by hhugo** [comment]
Thanks, should be released soon. https://github.com/ocaml/opam-repository/pull/23355

**Pull request merged**